src/net/stackerdb/db.rs

```diff
@@ -0,0 +1,602 @@
+        stackerdb_table(smart_contract)
+    );
+    let args: &[&dyn ToSql] = &[rc_consensus_hash, &chunk_id];
+    query_row(conn, &sql, args).map_err(|e| e.into())
+}
+
+impl<'a> StackerDBTx<'a> {
+    pub fn commit(self) -> Result<(), net_error> {
+        self.sql_tx.commit().map_err(net_error::from)
+    }
+
+    pub fn conn(&self) -> &DBConn {
+        &self.sql_tx
+    }
+
+    /// Create a stacker DB table and its indexes.
+    /// Idempotent.
+    pub fn create_stackerdb(&self, smart_contract: &ContractId) -> Result<(), net_error> {
+        let qry = format!(
+            "CREATE TABLE IF NOT EXISTS {}{}",
+            stackerdb_table(smart_contract),
+            CHUNKS_DB_TABLE_BODY
+        );
+        let mut stmt = self.sql_tx.prepare(&qry)?;
+        stmt.execute(NO_PARAMS)?;
+
+        for (index_name, index_body) in CHUNKS_DB_INDEX_BODIES.iter() {
+            let qry = format!(
+                "CREATE INDEX IF NOT EXISTS index_{}_{} ON {}{}",
+                stackerdb_table(smart_contract),
+                index_name,
+                stackerdb_table(smart_contract),
+                index_body
+            );
+            let mut stmt = self.sql_tx.prepare(&qry)?;
+            stmt.execute(NO_PARAMS)?;
+        }
+
+        let qry = "REPLACE INTO databases (table_name,smart_contract_id) VALUES (?1,?2)";
+        let args: &[&dyn ToSql] = &[&stackerdb_table(smart_contract), &smart_contract];
+        let mut stmt = self.sql_tx.prepare(&qry)?;
+        stmt.execute(args)?;
+
+        Ok(())
+    }
+
+    /// Delete a stacker DB table and its indexes.
+    /// Idempotent.
+    pub fn delete_stackerdb(&self, smart_contract: &ContractId) -> Result<(), net_error> {
+        let qry = "DELETE FROM databases WHERE table_name = ?1";
+        let mut stmt = self.sql_tx.prepare(&qry)?;
+        stmt.execute(&[&stackerdb_table(smart_contract)])?;
+
+        let qry = format!("DROP TABLE IF EXISTS {}", stackerdb_table(smart_contract));
+        let mut stmt = self.sql_tx.prepare(&qry)?;
+        stmt.execute(NO_PARAMS)?;
+
+        for (index_name, _) in CHUNKS_DB_INDEX_BODIES.iter() {
+            let qry = format!(
+                "DROP INDEX IF EXISTS index_{}_{}",
+                stackerdb_table(smart_contract),
+                index_name
+            );
+            let mut stmt = self.sql_tx.prepare(&qry)?;
+            stmt.execute(NO_PARAMS)?;
+        }
+
+        Ok(())
+    }
+
+    /// List all stacker DB smart contracts we have available
+    pub fn get_stackerdbs(&self) -> Result<Vec<ContractId>, net_error> {
+        let sql = "SELECT smart_contract_id FROM databases ORDER BY table_name";
+        query_rows(&self.conn(), sql, NO_PARAMS).map_err(|e| e.into())
+    }
+
+    /// Set up a database's storage slots.
+    /// The slots must be in a deterministic order, since they are used to determine the chunk ID
+    /// (and thus the key used to authenticate them)
```

Comment:
What is a storage slot? It seems like this method is just writing empty data into the table. Why is this necessary?

Reply:
It's done this way because the follow-on StackerDB state sync PR needs to be able to extract the version vector for all of the DB's slots.  But in order to do that, those slots either have to be instantiated, or we need some kind of sparse version vector calculation.  The latter requires one SQL query per slot (since if slot data does not exist, we'd need to synthesize it on the fly), whereas this only requires a single query which can use an index.